README.md
Note: Modified version to fix screen brightness OSD for T60p
Original files at http://savannah.nongnu.org/download/tpb/

---

tpb - program to use the IBM ThinkPad(tm) special keys

All the information used in this program was gathered by observing the device
/dev/nvram while pressing the buttons. This was done with the command

'watch --differences=cumulative -n 0 xxd /dev/nvram'

The meaning of the bits in nvram is documented in doc/nvram.txt

This program was originally developed on a Thinkpad T21 (2647-4AG).
It is known to work on 

	- A 30
	- A 31
	- R 30 (no hardware mixer)
	- R 31 (no hardware mixer)
	- R 32 (no hardware mixer)
	- R 40
	- S 30
	- T 20
	- T 23
	- T 30 (no hardware switching lcd/crt/both)
	- T 40
	- X 20
	- X 24
	- X 30
	- X 31
	- X 40

Use it at your own risk!!!

Installation:

- enable nvram in your kernel (section character devices), either as module
  or compiled into the kernel

- if you use nvram as module, load it

- make sure that the device /dev/nvram exists, else create it with
  'mknod /dev/nvram c 10 144'

- if you want to use on-screen display install xosd
  (http://www.ignavus.net/software.html)

- if you use the cvs version run "make -f Makefile.dist all"

- run the usual "./configure && make && make install"

- modify the template configuration file /usr/local/etc/tpbrc to change defaults

- run it with something like
  'tpb --osd=on --verbose --thinkpad="/usr/bin/X11/xterm -T ntpctl -e ntpctl"'

  NOTE: With tpb 0.6.2 or later you need to specify the full path for
        applications!

- have a look at the documentation in the man page, the template configuration
  file tpbrc and the output of 'tpb -h'


ChangeLog
2005-07-25 10:48  mbr

	* debian/changelog:

	chg: preparations for release.

2005-07-25 10:40  mbr

	* doc/tpbrc, man/tpb.1:

	chg: Minor changes to documentation.

2005-07-25 10:34  mbr

	* po/: de.po, tpb.pot:

	chg: Updated language files.

2005-07-18 16:15  mbr

	* configure.in, doc/tpbrc, man/tpb.1, src/cfg.c, src/cfg.h,
	  src/tpb.c, src/tpb.h:

	add: New config file options to selectivly turn on and off OSD messages.

2005-04-11 11:59  mbr

	* src/tpb.c:

	- change to users home directory before executing an application (thanks to
	  Florian Reitmeir)

2005-02-16 14:33  mbr

	* debian/changelog:

	chg: preparations for new Debian package release.

2005-02-16 14:17  mbr

	* debian/: README.Debian, tpb.devfs, tpb.postinst:

	chg: made /dev/nvram not group writeable by default.
	bug: closes Debian bug #288042

2005-02-16 13:41  mbr

	* man/tpb.1:

	add: Hint about the need for the complete path to executables.
	bug: closes Debian bug #277311

2005-02-16 13:20  mbr

	* debian/po/cs.po:

	add: Czech translation of tpb debconf messages. Thanks to Miroslav Kure!
	bug: closes Debian bug #294747

2004-08-31 16:08  mbr

	* README:

	add: X31 is also supported

2004-08-23 13:01  mbr

	* configure.in, debian/changelog, man/tpb.1:

	chg: preparations for release 0.6.3

2004-08-22 23:25  mbr

	* debian/: tpb.config, tpb.postinst:

	fix: udev could be recognized by /dev/udev.tdb NOT by /dev/udev/tdb

2004-08-22 22:48  mbr

	* debian/: tpb.postinst, tpb.xsession:

	chg: removed unneeded quotes

2004-08-22 22:45  mbr

	* debian/tpb.config:

	chg: fixed some issues pointed out by Sam Hocevar.

2004-08-22 22:42  mbr

	* debian/po/fr.po:

	chg: updated French translation. Thanks to Michel Grentzinger!

2004-08-22 15:45  mbr

	* src/tpb.c:

	chg: close the file descriptors the same way both times

2004-08-22 15:40  mbr

	* debian/po/de.po:

	chg: removed fuzzy and updated date

2004-08-22 15:29  mbr

	* debian/po/ja.po:

	chg: updated Japanese debconf texts. Thanks to Hideki Yamane!

2004-08-22 15:24  mbr

	* debian/README.Debian:

	add: added information about /dev/nvram, permissions and groups

2004-08-20 17:29  mbr

	* src/tpb.c:

	chg: close all file descriptors (except stdin, stdout, stderr) in forked
	     application, because nvram is after performance optimation allways open.

2004-08-20 12:58  mbr

	* src/tpb.c:

	chg: code police :)

2004-08-20 12:56  mbr

	* po/: de.po, tpb.pot:

	chg: synced template and German language with sources

2004-08-20 12:50  mbr

	* debian/tpb.postinst:

	chg: chmod should only turn on needed bits, but not change others.

2004-08-20 12:49  mbr

	* debian/: tpb.config, tpb.preinst:

	chg: only db_set variables if their value is not of zero length.
	chg: only ask the user for device creation if the device does not match the
	     requirements already.

2004-08-19 22:59  mbr

	* debian/: rules, tpb.config, tpb.postinst, tpb.postrm,
	  tpb.preinst, tpb.xsession:

	chg: moved startup configuration from /etc/X11/Xsession.d/90tpb to
	     /etc/default/tpb
	chg: reworked the debian scripts in general

2004-08-19 17:12  mbr

	* README:

	chg: indentation

2004-08-19 17:11  mbr

	* debian/: tpb.config, tpb.devfs, tpb.postinst, tpb.templates,
	  po/de.po, po/templates.pot:

	add: support for udev filesystem
	chg: adjusted German translation

2004-08-19 11:14  mbr

	* src/tpb.c:

	chg: check if thinkpad_state structure changed at all, before looking into
	     detail what changed. If nothing changed, just do nothing.

2004-08-18 16:57  mbr

	* src/tpb.c:

	chg: open nvram device only once and never close it in get_nvram_state().
	     The files will be closed a the end of the program.
	     This decreases the CPU load further. Now it is almost the same as not
	     running tpb at all.

2004-08-18 16:00  mbr

	* src/cfg.h:

	fix: now really use 200ms as default

2004-08-18 15:57  mbr

	* doc/tpbrc, man/tpb.1, src/cfg.c, src/cfg.h, src/tpb.c:

	chg: adjusted poll interval to 200ms, which has a acceptable responsiveness
	     and CPU load
	chg: read only needed bytes from nvram
	add: dynamic default display in the help screen

2004-08-18 13:22  mbr

	* debian/control:

	chg: fixed maintainer to match my Debian mail address

2004-08-18 12:47  mbr

	* debian/: README.Debian, changelog, control:

	chg: preparations for new Debian package release.

2004-08-18 12:38  mbr

	* CREDITS:

	add: forgot Soren S. Jorvang in CREDITS for his NetBSD support

2004-07-14 17:21  mbr

	* debian/tpb.xsession:

	chg: use = instead of == in test for strict POSIX compliance.

2004-05-19 16:26  mbr

	* man/tpb.1, debian/changelog:

	chg: preparations for release 0.6.2

2004-05-18 13:32  mbr

	* po/: da.po, de.po, ru.po, tpb.pot:

	chg: updated Russian translation (thanks to Grigory Batalov!), Danish
	     translation (thanks to Jes R. Lungskov) and German translation.

2004-05-18 13:21  mbr

	* src/tpb.c:

	fix: corrected typo programm -> program

2004-05-14 13:29  mbr

	* doc/callback_example.sh:

	add: wireless keypress event added to callback example.

2004-05-14 11:07  mbr

	* po/ru.po:

	chg: updated Russian translation. Many thanks to Grigory Batalov!

2004-05-14 10:55  mbr

	* po/: de.po, tpb.pot:

	chg: updated translation template and German translation

2004-05-11 13:20  mbr

	* man/tpb.1:

	fix: quoted some '-' signs to get hyphens instead of minus signs.

2004-05-07 14:23  mbr

	* tpb.spec:

	del: because the major rpm based distributions (Red Hat, Mandrake, Suse)
	     already include tpb and Dag Wieers has also an package and this file is
	     not supported it should be removed.

2004-05-05 12:25  mbr

	* README:

	add: new supported ThinkPad X40
	add: note about the need of fill path for applications

2004-04-29 17:45  mbr

	* doc/nvram.txt, src/tpb.c:

	fix: workaround for display state, because kernel 2.6 doesn't toggle bit 6 in
	     byte 0x57 of nvram anymore.

2004-04-29 13:12  mbr

	* configure.in, man/tpb.1, po/de.po, po/tpb.pot, src/cfg.c,
	  src/cfg.h, src/tpb.c, src/tpb.h:


config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.in by autoheader.  */

/* Define to 1 if using 'alloca.c'. */
/* #undef C_ALLOCA */

/* Define to 1 if translation of program messages to the user's native
   language is requested. */
#define ENABLE_NLS 1

/* Define to 1 if you have 'alloca', as a function or macro. */
#define HAVE_ALLOCA 1

/* Define to 1 if <alloca.h> works. */
#define HAVE_ALLOCA_H 1

/* Define to 1 if you have the <argz.h> header file. */
#define HAVE_ARGZ_H 1

/* Define to 1 if you have the `dcgettext' function. */
#define HAVE_DCGETTEXT 1

/* Define to 1 if you have the <fcntl.h> header file. */
#define HAVE_FCNTL_H 1

/* Define to 1 if you have the `feof_unlocked' function. */
#define HAVE_FEOF_UNLOCKED 1

/* Define to 1 if you have the `fgets_unlocked' function. */
#define HAVE_FGETS_UNLOCKED 1

/* Define to 1 if you have the `getcwd' function. */
#define HAVE_GETCWD 1

/* Define to 1 if you have the `getegid' function. */
#define HAVE_GETEGID 1

/* Define to 1 if you have the `geteuid' function. */
#define HAVE_GETEUID 1

/* Define to 1 if you have the `getgid' function. */
#define HAVE_GETGID 1

/* Define to 1 if you have the <getopt.h> header file. */
#define HAVE_GETOPT_H 1

/* Define to 1 if you have the `getpagesize' function. */
#define HAVE_GETPAGESIZE 1

/* Define if the GNU gettext() function is already present or preinstalled. */
#define HAVE_GETTEXT 1

/* Define to 1 if you have the `getuid' function. */
#define HAVE_GETUID 1

/* Define if you have the iconv() function. */
#define HAVE_ICONV 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define if you have <langinfo.h> and nl_langinfo(CODESET). */
#define HAVE_LANGINFO_CODESET 1

/* Define if your <locale.h> file defines LC_MESSAGES. */
#define HAVE_LC_MESSAGES 1

/* Define to 1 if you have the `X11' library (-lX11). */
#define HAVE_LIBX11 1

/* Defined to 1 if you have the xosd library */
#define HAVE_LIBXOSD 1

/* Define to version number of the xosd library. Numbering scheme: 700 >=
   0.7.0 ; 10000 >= 1.0.0 ; 20000 >= 2.0.0 ; 20200 >= 2.2.0 */
#define HAVE_LIBXOSD_VERSION 20200

/* Define to 1 if you have the <limits.h> header file. */
#define HAVE_LIMITS_H 1

/* Define to 1 if you have the <locale.h> header file. */
#define HAVE_LOCALE_H 1

/* Define to 1 if you have the <malloc.h> header file. */
#define HAVE_MALLOC_H 1

/* Define to 1 if you have the `mempcpy' function. */
#define HAVE_MEMPCPY 1

/* Define to 1 if you have a working `mmap' system call. */
/* #undef HAVE_MMAP */

/* Define to 1 if you have the `munmap' function. */
#define HAVE_MUNMAP 1

/* Define to 1 if you have the <nl_types.h> header file. */
#define HAVE_NL_TYPES_H 1

/* Define to 1 if you have the `putenv' function. */
#define HAVE_PUTENV 1

/* Define to 1 if you have the `setenv' function. */
#define HAVE_SETENV 1

/* Define to 1 if you have the `setlocale' function. */
#define HAVE_SETLOCALE 1

/* Define to 1 if you have the <stddef.h> header file. */
#define HAVE_STDDEF_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdio.h> header file. */
#define HAVE_STDIO_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the `stpcpy' function. */
#define HAVE_STPCPY 1

/* Define to 1 if you have the `strcasecmp' function. */
#define HAVE_STRCASECMP 1

/* Define to 1 if you have the `strchr' function. */
#define HAVE_STRCHR 1

/* Define to 1 if you have the `strdup' function. */
#define HAVE_STRDUP 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the `strtoul' function. */
#define HAVE_STRTOUL 1

/* Define to 1 if you have the <sys/param.h> header file. */
#define HAVE_SYS_PARAM_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the `tsearch' function. */
#define HAVE_TSEARCH 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if you have the <wait.h> header file. */
#define HAVE_WAIT_H 1

/* Define to 1 if you have the `__argz_count' function. */
#define HAVE___ARGZ_COUNT 1

/* Define to 1 if you have the `__argz_next' function. */
#define HAVE___ARGZ_NEXT 1

/* Define to 1 if you have the `__argz_stringify' function. */
#define HAVE___ARGZ_STRINGIFY 1

/* Define as const if the declaration of iconv() needs const. */
#define ICONV_CONST 

/* Name of package */
#define PACKAGE "tpb"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME ""

/* Define to the full name and version of this package. */
#define PACKAGE_STRING ""

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME ""

/* Define to the home page for this package. */
#define PACKAGE_URL ""

/* Define to the version of this package. */
#define PACKAGE_VERSION ""

/* If using the C implementation of alloca, define if you know the
   direction of stack growth for your system; otherwise it will be
   automatically deduced at runtime.
	STACK_DIRECTION > 0 => grows toward higher addresses
	STACK_DIRECTION < 0 => grows toward lower addresses
	STACK_DIRECTION = 0 => direction of growth unknown */
/* #undef STACK_DIRECTION */

/* Define to 1 if all of the C90 standard headers exist (not just the ones
   required in a freestanding environment). This macro is provided for
   backward compatibility; new code need not use it. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "0.6.4"

/* Define to 1 if the X Window System is missing or not being used. */
/* #undef X_DISPLAY_MISSING */

/* Define to empty if `const' does not conform to ANSI C. */
/* #undef const */

/* Define to `__inline__' or `__inline' if that's what the C compiler
   calls it, or to nothing if 'inline' is not supported under any name.  */
#ifndef __cplusplus
/* #undef inline */
#endif

/* Define to `long int' if <sys/types.h> does not define. */
/* #undef off_t */

/* Define to `unsigned int' if <sys/types.h> does not define. */
/* #undef size_t */


src/tpb.h
/* tpb - program to use the IBM ThinkPad(tm) special keys
 * Copyright (C) 2002-2005 Markus Braun <markus.braun@krawel.de>
 *
 * This file is part of tpb.
 *
 * tpb is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * tpb is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with tpb; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef __TPB_H__
#define __TPB_H__

#if ENABLE_NLS
#include <libintl.h>
#define _(String) gettext (String)
#else /* ENABLE_NLS */
#define _(String) (String)
#endif /* ENABLE_NLS */

#define MAX_VOLUME 100
#define CALLBACK_CMD_LENGTH 256
#define CALLBACK_CMD_ARGS 30

#define KEYCODE_HOME 178
#define KEYCODE_SEARCH 229
#define KEYCODE_MAIL 236
#define KEYCODE_FAVORITES 230
#define KEYCODE_RELOAD 231
#define KEYCODE_ABORT 232
#define KEYCODE_BACKWARD 234
#define KEYCODE_FORWARD 233
#define KEYCODE_FN_KEY 227

/* all nvram toggle values are stored in bit 0 */
/* all xevents toggle values are stored in bit 1 */
typedef struct {
  unsigned int thinkpad_toggle;     /* ThinkPad button */
  unsigned int zoom_toggle;         /* zoom toggle */
  unsigned int display_toggle;      /* display toggle */
  unsigned int home_toggle;         /* Home button */
  unsigned int search_toggle;       /* Search button */
  unsigned int mail_toggle;         /* Mail button */
  unsigned int wireless_toggle;     /* Wireless button */
  unsigned int favorites_toggle;    /* Favorites button */
  unsigned int reload_toggle;       /* Reload button */
  unsigned int abort_toggle;        /* Abort button */
  unsigned int backward_toggle;     /* Backward button */
  unsigned int forward_toggle;      /* Forward button */
  unsigned int fn_toggle;           /* Fn button */
  unsigned int thinklight_toggle;   /* ThinkLight */
  unsigned int hibernate_toggle;    /* hibernation/suspend toggle */
  unsigned int display_state;       /* display state */
  unsigned int expand_toggle;       /* hv expansion state */
  unsigned int brightness_level;    /* brightness level */
  unsigned int brightness_toggle;   /* brightness toggle */
  unsigned int volume_level;        /* volume level */
  unsigned int volume_toggle;       /* volume toggle */
  unsigned int mute_toggle;         /* mute toggle */
  unsigned int ac_state;            /* ac connected */
  unsigned int powermgt_ac;         /* power management mode ac */
  unsigned int powermgt_battery;    /* power management mode battery */
} t_thinkpad_state;

#endif /* __TPB_H__*/

/* vim600:set fen:set fdm=marker:set fdl=0: */


src/cfg.h
/* tpb - program to use the IBM ThinkPad(tm) special keys
 * Copyright (C) 2002-2005 Markus Braun <markus.braun@krawel.de>
 *
 * This file is part of tpb.
 *
 * tpb is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * tpb is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with tpb; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef __CFG_H__
#define __CFG_H__

#if ENABLE_NLS
#include <libintl.h>
#define _(String) gettext (String)
#else /* ENABLE_NLS */
#define _(String) (String)
#endif /* ENABLE_NLS */

enum state {UNDEFINED = -1, STATE_OFF, STATE_ON, STATE_AUTO};
enum mode {MODE_COMMENT, MODE_INDENT, MODE_KEYWORD, MODE_SEPARATOR, MODE_ARGUMENT, MODE_END};

/* configuration defaults */
#define DEFAULT_DAEMON STATE_OFF
#define DEFAULT_APM STATE_OFF
#define DEFAULT_POWERMGT STATE_AUTO
#define DEFAULT_XEVENTS STATE_ON
#define DEFAULT_VERBOSE STATE_OFF
#define DEFAULT_POLLTIME 200000
#define DEFAULT_NVRAMDEV "/dev/nvram"
#define DEFAULT_NVRAMDEV_DEVFS "/dev/misc/nvram"
#define DEFAULT_MIXER STATE_OFF
#define DEFAULT_MIXERSTEPS 14
#define DEFAULT_MIXERDEV "/dev/mixer"
#ifdef HAVE_LIBXOSD
#define DEFAULT_OSD STATE_ON
#define DEFAULT_OSDZOOM UNDEFINED
#define DEFAULT_OSDTHINKLIGHT UNDEFINED
#define DEFAULT_OSDDISPLAY UNDEFINED
#define DEFAULT_OSDHVEXPANSION UNDEFINED
#define DEFAULT_OSDBRIGHTNESS UNDEFINED
#define DEFAULT_OSDVOLUME UNDEFINED
#define DEFAULT_OSDPOWERMGT UNDEFINED
#define DEFAULT_OSDLINES 2
#if HAVE_LIBXOSD_VERSION > 700
#define DEFAULT_OSDFONT osd_default_font
#else
#define DEFAULT_OSDFONT "10x20"
#endif /* HAVE_LIBXOSD_VERSION */
#define DEFAULT_OSDCOLOR "BLUE"
#define DEFAULT_OSDTIMEOUT 3
#define DEFAULT_OSDSHADOW 2
#define DEFAULT_OSDSHADOWCOLOR "BLACK"
#define DEFAULT_OSDOUTLINE 1
#define DEFAULT_OSDOUTLINECOLOR "BLACK"
#define DEFAULT_OSDVERTICAL 25
#define DEFAULT_OSDHORIZONTAL 25
#define DEFAULT_OSDPOS XOSD_bottom
#define DEFAULT_OSDALIGN XOSD_left
#endif /* HAVE_LIBXOSD */

/* configuration files */
#ifdef SYSCONFDIR
#define GLOBAL_CONFIG_FILE SYSCONFDIR "/tpbrc"
#else /* SYSCONFDIR */
#define GLOBAL_CONFIG_FILE "/etc/tpbrc"
#endif /* SYSCONFDIR */
#define PRIVATE_CONFIG_FILE ".tpbrc"

/* Keywords in configuration file */
#define CFG_APM "apm"
#define CFG_APM_ON "on"
#define CFG_APM_OFF "off"
#define CFG_POWERMGT "powermgt"
#define CFG_POWERMGT_ON "on"
#define CFG_POWERMGT_OFF "off"
#define CFG_POWERMGT_AUTO "auto"
#define CFG_XEVENTS "xevents"
#define CFG_XEVENTS_ON "on"
#define CFG_XEVENTS_OFF "off"
#define CFG_NVRAM "nvram"
#define CFG_POLLTIME "polltime"
#define CFG_THINKPAD "thinkpad"
#define CFG_HOME "home"
#define CFG_SEARCH "search"
#define CFG_MAIL "mail"
#define CFG_WIRELESS "wireless"
#define CFG_FAVORITES "favorites"
#define CFG_RELOAD "reload"
#define CFG_ABORT "abort"
#define CFG_BACKWARD "backward"
#define CFG_FORWARD "forward"
#define CFG_FN "fn"
#define CFG_CALLBACK "callback"
#define CFG_MIXER "mixer"
#define CFG_MIXER_ON "on"
#define CFG_MIXER_OFF "off"
#define CFG_MIXERSTEPS "mixersteps"
#define CFG_MIXERDEV "mixerdevice"
#ifdef HAVE_LIBXOSD
#define CFG_OSDFONT "osdfont"
#define CFG_OSDCOLOR "osdcolor"
#define CFG_OSDTIMEOUT "osdtimeout"
#define CFG_OSDOFFSET "osdoffset"
#define CFG_OSDSHADOW "osdshadow"
#if HAVE_LIBXOSD_VERSION >= 20200
#define CFG_OSDSHADOWCOLOR "osdshadowcolor"
#define CFG_OSDOUTLINE "osdoutline"
#define CFG_OSDOUTLINECOLOR "osdoutlinecolor"
#endif /* HAVE_LIBXOSD_VERSION */
#define CFG_OSDVERTICAL "osdvertical"
#define CFG_OSDHORIZONTAL "osdhorizontal"
#define CFG_OSDPOS "osdpos"
#define CFG_OSDPOS_TOP "top"
#define CFG_OSDPOS_MIDDLE "middle"
#define CFG_OSDPOS_BOTTOM "bottom"
#define CFG_OSDALIGN "osdalign"
#define CFG_OSDALIGN_LEFT "left"
#define CFG_OSDALIGN_CENTER "center"
#define CFG_OSDALIGN_RIGHT "right"
#define CFG_OSD "osd"
#define CFG_OSD_ON "on"
#define CFG_OSD_OFF "off"
#define CFG_OSDZOOM "osdzoom"
#define CFG_OSDZOOM_ON "on"
#define CFG_OSDZOOM_OFF "off"
#define CFG_OSDTHINKLIGHT "osdthinklight"
#define CFG_OSDTHINKLIGHT_ON "on"
#define CFG_OSDTHINKLIGHT_OFF "off"
#define CFG_OSDDISPLAY "osddisplay"
#define CFG_OSDDISPLAY_ON "on"
#define CFG_OSDDISPLAY_OFF "off"
#define CFG_OSDHVEXPANSION "osdhvexpansion"
#define CFG_OSDHVEXPANSION_ON "on"
#define CFG_OSDHVEXPANSION_OFF "off"
#define CFG_OSDBRIGHTNESS "osdbrightness"
#define CFG_OSDBRIGHTNESS_ON "on"
#define CFG_OSDBRIGHTNESS_OFF "off"
#define CFG_OSDVOLUME "osdvolume"
#define CFG_OSDVOLUME_ON "on"
#define CFG_OSDVOLUME_OFF "off"
#define CFG_OSDPOWERMGT "osdpowermgt"
#define CFG_OSDPOWERMGT_ON "on"
#define CFG_OSDPOWERMGT_OFF "off"
#endif /* HAVE_LIBXOSD */

#define BUFFER_SIZE 256

typedef struct {
  int daemon;
  int apm;
  int powermgt;
  int xevents;
  char *nvram;
  int polltime;
  char *tpb_cmd;
  char *home_cmd;
  char *search_cmd;
  char *mail_cmd;
  char *wireless_cmd;
  char *favorites_cmd;
  char *reload_cmd;
  char *abort_cmd;
  char *backward_cmd;
  char *forward_cmd;
  char *fn_cmd;
  char *callback;
  int mixer;
  int mixersteps;
  char *mixerdev;
  int verbose;
#ifdef HAVE_LIBXOSD
  int osd;
  int osdzoom;
  int osdthinklight;
  int osddisplay;
  int osdhvexpansion;
  int osdbrightness;
  int osdvolume;
  int osdpowermgt;
  char *osdfont;
  char *osdcolor;
  int osdtimeout;
  int osdshadow;
#if HAVE_LIBXOSD_VERSION >= 20200
  char *osdshadowcolor;
  int osdoutline;
  char *osdoutlinecolor;
#endif /* HAVE_LIBXOSD_VERSION */
  int osdvertical;
  int osdhorizontal;
  xosd_pos osdpos;
#if HAVE_LIBXOSD_VERSION >= 10000
  xosd_align osdalign;
#endif /* HAVE_LIBXOSD_VERSION */
#endif /* HAVE_LIBXOSD */
} config;

int parse_cfg_file(config *cfg, char *name);
void parse_option(config *cfg, int argc, char **argv);
void init_cfg(config *cfg);
void clear_cfg(config *cfg);
void free_cfg(config *cfg);
void override_cfg(config *master, config *slave);
void set_value(char *key, char *arg, config *cfg);
void print_usage(const char *prog);

#endif /* __CFG_H__ */

/* vim600:set fen:set fdm=marker:set fdl=0: */
